lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,543 @@
+
+    #[test]
+    #[allow(clippy::too_many_lines)]
+    fn test_remove_at_path() {
+        struct TestCase {
+            kind: Kind,
+            path: LookupBuf,
+            returned: Option<Kind>,
+            mutated: Kind,
+        }
+
+        for (
+            title,
+            TestCase {
+                mut kind,
+                path,
+                returned,
+                mutated,
+            },
+        ) in HashMap::from([
+            (
+                "primitive",
+                TestCase {
+                    kind: Kind::bytes(),
+                    path: "foo".into(),
+                    returned: None,
+                    mutated: Kind::bytes(),
+                },
+            ),
+            (
+                "multiple primitives",
+                TestCase {
+                    kind: Kind::integer().or_regex(),
+                    path: "foo".into(),
+                    returned: None,
+                    mutated: Kind::integer().or_regex(),
+                },
+            ),
+            (
+                "object w/ matching path",
+                TestCase {
+                    kind: Kind::object(BTreeMap::from([("foo".into(), Kind::integer())])),
+                    path: "foo".into(),
+                    returned: Some(Kind::integer()),
+                    mutated: Kind::object(BTreeMap::default()),
+                },
+            ),
+            (
+                "object w/o matching path",
+                TestCase {
+                    kind: Kind::object(BTreeMap::from([("foo".into(), Kind::integer())])),
+                    path: "bar".into(),
+                    returned: None,
+                    mutated: Kind::object(BTreeMap::from([("foo".into(), Kind::integer())])),
+                },
+            ),
+            (
+                "array w/ matching path",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[1]").unwrap(),
+                    returned: Some(Kind::integer()),
+                    mutated: Kind::array(BTreeMap::default()),
+                },
+            ),
+            (
+                "array w/o matching path",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[2]").unwrap(),
+                    returned: None,
+                    mutated: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                },
+            ),
+            (
+                "array w/ negative indexing",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[-1]").unwrap(),
+                    returned: None,
+                    mutated: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                },
+            ),
```

Comment:
It's not a programming error per sé, there are places where people can use negative indexing in paths. But yeah, it is an edge-case, and one we (currently) don't really support.

I can see an update to the type system in the future that would allow us to better track array sizes, and then allowing or rejecting negative indexing more explicitly.